### PR TITLE
Feature/init array default values

### LINKED ISSUE #291: initialize array with default values
feature request:
initialize mutable array of size with default values.
at the moment: `emptyArray[Int](5)` creates `( , , , , ,)` of unknown type, in js probably null
requested behaviour: `newArray[T](length, defaultValue)`
 `newArray[Int](5,0)` creates: `(0,0,0,0,0)`

### PR DESCRIPTION
**Pull request by IR0NSIGHT**
allows initializing a mutable array of size n with default values 
skips having to init an array of size n with unknown type values and then having to fill the array with values.

added some QOL features such as init array from a function by index and init array from hardcoded list

fixes #291

### CODE REVIEW

**Comment by IR0NSIGHT** [comment]
@phischu

File: examples/pos/arrays.effekt [hunk 1]
```diff
@@ -17,4 +17,13 @@ def main() = {
     println(show(get(arr, 4)));
     put(arr, 4, "foo");
     println(show(get(arr, 4)))
+
+    val arrWithVals = initArray(4,"Hello World");
+    println(arrWithVals)
+
+    val arrByF = arrayFromLoop(5){ i => i};
```
Comment (phischu) [inline examples/pos/arrays.effekt:24]: Please add a space `{ i => i }`

File: libraries/js/immutable/list.effekt [hunk 2]
```diff
@@ -20,6 +20,14 @@ def foreach[A](l: List[A]) { f: A => Unit } : Unit = {
   }
 }
 
+def foreachIndex[T](list: List[T], i: Int){ f: (Int, T) => Unit}: Unit = {
```
Comment (phischu) [inline libraries/js/immutable/list.effekt:23]: Make a local loop (tail-called function) so users don't have to pass in the initial index

File: libraries/js/mutable/array.effekt [hunk 3]
```diff
@@ -8,6 +8,29 @@ extern type Array[T]
 
 def emptyArray[T](): Array[T] = emptyArray(0)
 
+//initialize array with default values
+def initArray[T](size: Int, default: T) : Array[T] = {
```
Comment (phischu) [inline libraries/js/mutable/array.effekt:12]: Could we call it `filledArray` or `arrayFromValue` in accordance with the other two?

**Review by phischu — COMMENTED** [review]
Very good job

File: libraries/js/mutable/array.effekt (hunk 3 above)
Comment (IR0NSIGHT) [reply libraries/js/mutable/array.effekt:12]: done

File: libraries/js/immutable/list.effekt (hunk 2 above)
Comment (IR0NSIGHT) [reply libraries/js/immutable/list.effekt]: done

File: examples/pos/arrays.effekt (hunk 1 above)
Comment (IR0NSIGHT) [reply examples/pos/arrays.effekt:24]: :eyes:

**Comment by IR0NSIGHT** [comment]
@phischu done

File: libraries/js/mutable/array.effekt [hunk 4]
```diff
@@ -8,6 +8,29 @@ extern type Array[T]
...
+    val arr = emptyArray[T](size)
+    each(0,arr.size()){ i=> put(arr, i, default)}
+    return arr;
+}
+
+def arrayFromLoop[T](size: Int) { index: Int => T}: Array[T] = {
+    val arr = emptyArray[T](size)
+    each(0,arr.size()){ i=> put(arr, i, index(i))}
+    return arr
+}
```
Comment (phischu) [inline libraries/js/mutable/array.effekt:22]:
```suggestion
def arrayFromValue[T](size: Int, default: T) : Array[T] = {
  val arr = emptyArray[T](size)
  each(0,arr.size()){ i=> put(arr, i, default)}
  return arr;
}

def arrayFromLoop[T](size: Int) { index: Int => T}: Array[T] = {
  val arr = emptyArray[T](size)
  each(0,arr.size()){ i=> put(arr, i, index(i))}
  return arr
}
```

**Review by phischu — APPROVED** [review]

File: libraries/js/immutable/list.effekt [hunk 5]
```diff
@@ -20,6 +20,17 @@ def foreach[A](l: List[A]) { f: A => Unit } : Unit = {
   }
 }
 
+def foreachIndex[T](list: List[T]){ f: (Int, T) => Unit}: Unit = {
+  def worker[T](list: List[T], i: Int){ f: (Int, T) => Unit}: Unit = {
+    list match {
+      case Nil() => ();
+      case Cons(head, tail) => f(i, head); worker(tail, i+1){f}
```
Comment (b-studios) [inline libraries/js/immutable/list.effekt:27]: @IR0NSIGHT @phischu `f` is stable throughout the recursion, please consider dropping it from worker.

**Pull request merged**